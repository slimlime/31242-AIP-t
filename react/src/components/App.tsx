import React from "react";
import "../assets/css/App.css";
import {
  Switch,
  Route,
  BrowserRouter as Router,
  Redirect,
} from "react-router-dom";
import Login from "../pages/login";
import SignUp from "../pages/signup";
import UserProfile from "../pages/userprofile";
import Home from "../pages/home";
import { UserContext } from "./user-context";

type AppState = {
  user: Object;
  updateUser: (newUser: Object) => void;
};

class App extends React.Component {
  state: AppState = {
    user: window.localStorage.getItem("user") ?? { name: "?" },
    updateUser: (newUser: Object) => {
      this.setState({ user: newUser });
      // Caches current authenticated user to local storage for offline viewing
      window.localStorage.setItem("user", JSON.stringify(newUser));
    },
  };

  render() {
    return (
      <Router>
        <UserContext.Provider value={this.state}>
          <Switch>
            <Route path="/user" component={UserProfile} />
            <Route path="/signup" component={SignUp} />
            <Route path="/login" component={Login} />
            <Route path="/home" component={Home} />
            <Route exact path="/">
              <Redirect to="/home" />
            </Route>
          </Switch>
        </UserContext.Provider>
      </Router>
    );
  }
}

export default App;
