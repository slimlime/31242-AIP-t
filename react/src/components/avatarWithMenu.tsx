import React from "react";
import { Link as RouterLink } from "react-router-dom";
import "../assets/css/avatarWithMenu.css";
import {
  Avatar,
  Link,
  ListItemIcon,
  Menu,
  MenuItem,
  MenuProps,
  withStyles,
  ListItemText,
} from "@material-ui/core";
import {
  ThumbsUpDown,
  ExitToApp,
  AssignmentInd,
  PersonAdd,
  AccountCircle,
} from "@material-ui/icons";
import { UserContext } from "./user-context";

// kudos to https://material-ui.com/components/menus/
const StyledMenu = withStyles({
  paper: {
    border: "1px solid #d3d4d5",
  },
})((props: MenuProps) => (
  <Menu
    elevation={0}
    getContentAnchorEl={null}
    anchorOrigin={{
      vertical: "bottom",
      horizontal: "center",
    }}
    transformOrigin={{
      vertical: "top",
      horizontal: "center",
    }}
    {...props}
  />
));

// kudos to https://material-ui.com/components/menus/
const StyledMenuItem = withStyles((theme) => ({
  root: {
    "&:focus": {
      backgroundColor: theme.palette.primary.main,
      "& .MuiListItemIcon-root, & .MuiListItemText-primary": {
        color: theme.palette.common.white,
      },
    },
  },
}))(MenuItem);

type AvatarWithMenuState = {
  userMenu: boolean;
  anchorEl: HTMLElement | null;
};

// most readable solution from https://stackoverflow.com/questions/33076177/getting-name-initials-using-js
function nameToUpperInitials(fullName: string) {
  const namesArray = fullName?.split(" ") ?? ["?"];
  if (namesArray.length === 1)
    return `${namesArray[0].charAt(0).toUpperCase()}`;
  else
    return `${namesArray[0].charAt(0).toUpperCase()}${namesArray[
      namesArray.length - 1
    ]
      .charAt(0)
      .toUpperCase()}`;
}

class AvatarWithMenu extends React.Component {
  state: AvatarWithMenuState = {
    userMenu: false,
    anchorEl: null,
  };

  static contextType = UserContext;

  render() {
    return (
      <div>
        {/* If user is logged in */}
        {this.context.user.name !== "?" && (
          <div>
            <Avatar
              onClick={(event: React.MouseEvent<HTMLElement>) => {
                this.setState({
                  userMenu: true,
                  anchorEl: event.currentTarget,
                });
              }}
              id="avatar"
            >
              {nameToUpperInitials(this.context.user.name)}
            </Avatar>
            <StyledMenu
              id="customized-menu"
              anchorEl={this.state.anchorEl}
              keepMounted
              open={this.state.userMenu}
              onClose={() => {
                this.setState({
                  userMenu: false,
                  anchorEl: null,
                });
              }}
            >
              <Link
                id="favoursLink"
                color="inherit"
                component={RouterLink}
                to={{ pathname: "/user", state: { tabIndex: 1 } }}
                style={{ textDecoration: "none" }}
              >
                <StyledMenuItem>
                  <ListItemIcon>
                    <ThumbsUpDown fontSize="small" />
                  </ListItemIcon>
                  <ListItemText primary="My Favours" />
                </StyledMenuItem>
              </Link>
              <MenuItem />
              <Link
                id="logOutLink"
                color="inherit"
                component={RouterLink}
                to="/home"
                // Clear user session
                onClick={() => {
                  this.context.updateUser({ name: "?" });
                  this.setState({ userMenu: false });
                  fetch(`/api/logout`, {
                    method: "GET",
                  }).catch((exception) => {
                    console.error("Error:", exception);
                  });
                }}
                style={{ textDecoration: "none" }}
              >
                <StyledMenuItem>
                  <ListItemIcon>
                    <ExitToApp fontSize="small" />
                  </ListItemIcon>
                  <ListItemText primary="Log Out" />
                </StyledMenuItem>
              </Link>
            </StyledMenu>
          </div>
        )}
        {/* User is not logged in */}
        {this.context.user.name === "?" && (
          <div>
            <Avatar
              onClick={(event: React.MouseEvent<HTMLElement>) => {
                this.setState({
                  userMenu: true,
                  anchorEl: event.currentTarget,
                });
              }}
              id="avatar"
            >
              <AccountCircle id="accountCircle" />
            </Avatar>
            <StyledMenu
              id="customized-menu"
              anchorEl={this.state.anchorEl}
              keepMounted
              open={this.state.userMenu}
              onClose={() => {
                this.setState({
                  userMenu: false,
                  anchorEl: null,
                });
              }}
            >
              <Link
                id="favoursLink"
                color="inherit"
                component={RouterLink}
                to={{ pathname: "/login", state: { tabIndex: 1 } }}
                style={{ textDecoration: "none" }}
              >
                <StyledMenuItem>
                  <ListItemIcon>
                    <AssignmentInd fontSize="small" />
                  </ListItemIcon>
                  <ListItemText primary="Login" />
                </StyledMenuItem>
              </Link>
              <Link
                id="requestLink"
                color="inherit"
                component={RouterLink}
                to={{ pathname: "/signup", state: { tabIndex: 2 } }}
                style={{ textDecoration: "none" }}
              >
                <StyledMenuItem>
                  <ListItemIcon>
                    <PersonAdd fontSize="small" />
                  </ListItemIcon>
                  <ListItemText primary="Sign Up" />
                </StyledMenuItem>
              </Link>
            </StyledMenu>
          </div>
        )}
      </div>
    );
  }
}

export { AvatarWithMenu, nameToUpperInitials };
