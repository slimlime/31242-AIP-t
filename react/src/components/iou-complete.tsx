import React from "react";
import { Redirect } from "react-router-dom";
import {
  Checkbox,
  Dialog,
  Grow,
  DialogTitle,
  Typography,
  DialogContent,
  DialogContentText,
  DialogActions,
  Button,
  Snackbar,
  Divider,
  TableContainer,
  Table,
  TableBody,
  TableRow,
  TableCell,
  Popover,
} from "@material-ui/core";
import CheckCircleOutlineIcon from "@material-ui/icons/CheckCircleOutline";
import RadioButtonUncheckedIcon from "@material-ui/icons/RadioButtonUnchecked";
import "../assets/css/iou-request.css";
import IouFavour from "./iou-single-favour";
import {
  requestEndpoint,
  iouOweEndpoint,
  iouOwedEndpoint,
} from "../api/endpoints";

type Item = {
  id: string;
  display_name: string;
};

type RewardItem = {
  id: string; // UUID;
  giver: { username: string; display_name: string };
  item: { id: string; display_name: string };
};

type IouCompleteProps = {
  id: string;
  is_completed: boolean;
  author: string;
  completed_by: string;
  claimed_time: string | null;
  created_time: string;
  rewards: RewardItem[];
  details: string;
  iouType: number;
  refreshTable: () => void;
};

type IouCompleteState = {
  completeIOU: boolean;
  submittedProof: any;
  snackMessage: string;
  completeSnack: boolean;
  AnchorEl: HTMLElement | null;
  unauthRep: boolean;
};

class IouComplete extends React.Component<IouCompleteProps, IouCompleteState> {
  private tempRewardDisplayName: string; // Not in state because it'll cause everything to go into a render loop
  constructor(props: IouCompleteProps) {
    super(props);
    this.tempRewardDisplayName = "";
  }

  state: IouCompleteState = {
    completeIOU: false,
    submittedProof: null,
    snackMessage: "",
    completeSnack: false,
    AnchorEl: null,
    unauthRep: false,
  };

  completeIouOwe() {
    const formData = new FormData();
    formData.append("proof", this.state.submittedProof);

    fetch(`${iouOweEndpoint}/${this.props.id}/complete`, {
      method: "PUT",
      body: formData,
    })
      .then((res) => {
        if (res.status === 200) {
          // Successful login 200
          this.setState({
            snackMessage: "IOU Proof Submitted & IOU Completed!",
          });
          this.setState({ completeSnack: true });
          this.props.refreshTable();
        } else if (res.status === 401) {
          this.setState({ unauthRep: true });
        } else {
          // Unsuccessful login (400)
          res.json().then((body) =>
            this.setState({
              snackMessage: body.errors,
              completeSnack: true,
            })
          );
        }
      })
      .catch((exception) => {
        console.log("Error:", exception);
      });
  }

  completeIouOwed() {
    fetch(`${iouOwedEndpoint}/${this.props.id}/complete`, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
    })
      .then((res) => {
        if (res.status === 200) {
          // Successful login 200
          this.setState({ snackMessage: "IOU Completed!" });
          this.setState({ completeSnack: true });
          this.props.refreshTable();
        } else if (res.status === 401) {
          this.setState({ unauthRep: true });
        } else {
          // Unsuccessful login (400)
          res.json().then((body) =>
            this.setState({
              snackMessage: body.errors,
              completeSnack: true,
            })
          );
        }
      })
      .catch((exception) => {
        console.log("Error:", exception);
      });
  }

  completeRequest() {
    const formData = new FormData();
    formData.append("proof", this.state.submittedProof);

    fetch(`${requestEndpoint}/${this.props.id}/complete`, {
      method: "PUT",
      body: formData,
    })
      .then((res) => {
        if (res.status === 200) {
          // Successful login 200
          this.setState({ snackMessage: "Request Completed!" });
          this.setState({ completeSnack: true });
          this.props.refreshTable();
        } else if (res.status === 401) {
          this.setState({ unauthRep: true });
        } else {
          // Unsuccessful login (400)
          res.json().then((body) =>
            this.setState({
              snackMessage: body.errors,
              completeSnack: true,
            })
          );
        }
      })
      .catch((exception) => {
        console.log("Error:", exception);
      });
  }

  fileContent() {
    if (this.state.submittedProof) {
      return (
        <div id="completeProofFileInfo">
          <DialogContentText variant="body2">
            {"Image Preview: "}
          </DialogContentText>
          <img
            src={URL.createObjectURL(this.state.submittedProof)}
            alt={this.state.submittedProof.name}
            id="completeProofImage"
          />
        </div>
      );
    } else {
      return (
        <DialogContentText variant="body2" id="completeProofFileInfo">
          {"Upload an image (JPEG/PNG) before pressing the 'Submit' button"}
        </DialogContentText>
      );
    }
  }

  renderPopUpRewards(reward: RewardItem) {
    return (
      <IouFavour
        key={reward.item.id}
        giverDisplayName={reward.giver.display_name}
        recieverDisplayName={"?"}
        item={reward.item}
      />
    );
  }

  checkSubmitButton(): boolean {
    if (this.props.iouType !== 0) {
      return !this.state.submittedProof;
    }
    return false;
  }

  render() {
    if (this.state.unauthRep) {
      return (
        <Redirect
          to={{
            pathname: "/login",
            state: {
              // This is just a pop up message on login page to signify session expired
              unauthenticated:
                "Your session has expired! Please sign in again :)",
            },
          }}
        />
      );
    }

    return (
      <div id="completeRequestItem">
        <div
          onClick={() => {
            if (!this.props.is_completed) {
              this.setState({ completeIOU: true });
            }
          }}
          className={this.props.is_completed ? "" : "cursorPointer"}
        >
          <Checkbox
            checked={this.props.is_completed}
            color="primary"
            disabled={true}
            checkedIcon={
              <CheckCircleOutlineIcon color="primary" id="completeIconSize" />
            }
            icon={
              <RadioButtonUncheckedIcon color="primary" id="completeIconSize" />
            }
          />
          {this.props.is_completed && (
            <div id="IouTaskComplete">
              {this.props.completed_by.length >= 7 ? (
                <Typography
                  id="taskCompleter"
                  className="cursorPointer"
                  onClick={(event: React.MouseEvent<HTMLElement>) =>
                    this.setState({ AnchorEl: event.currentTarget })
                  }
                >
                  {this.props.completed_by}
                </Typography>
              ) : (
                <div>
                  <Typography id="taskCompleter">
                    {this.props.completed_by}
                  </Typography>
                  <Typography id="taskTimeStamp">
                    {this.props.claimed_time}
                  </Typography>
                </div>
              )}
              <Popover
                id="popUpMargin"
                open={Boolean(this.state.AnchorEl)}
                anchorEl={this.state.AnchorEl}
                onClose={() => this.setState({ AnchorEl: null })}
                anchorOrigin={{
                  vertical: "bottom",
                  horizontal: "center",
                }}
                transformOrigin={{
                  vertical: "top",
                  horizontal: "center",
                }}
              >
                <Typography id="taskDetailPopUp">
                  {"Full Display Name"}
                  <Divider id="taskPopUpDivider" />
                  {this.props.completed_by}
                </Typography>
              </Popover>
            </div>
          )}
        </div>
        <Dialog
          maxWidth="xs"
          fullWidth={true}
          scroll="body"
          open={this.state.completeIOU}
          TransitionComponent={Grow}
          onClose={() =>
            this.setState({ completeIOU: false, submittedProof: null })
          }
          id="completeForm"
        >
          <DialogTitle disableTypography={true}>
            <Typography variant="h5" id="modalTitle">
              {"IOU Proof Submission"}
            </Typography>
          </DialogTitle>
          <Divider />
          <DialogContent className="content" id="completeDetailsTable">
            <DialogContentText variant="h6">{"IOU Details"}</DialogContentText>
            <TableContainer id="infoTable">
              <Table>
                <TableBody>
                  <TableRow>
                    <TableCell
                      align="center"
                      className="infoTableCell"
                      id="infoTableTitle"
                    >
                      {this.props.iouType === 2 ? "Request" : "IOU"} created by:
                    </TableCell>
                    <TableCell
                      align="center"
                      className="infoTableCell"
                      id="infoTableContent"
                    >
                      {this.props.author}
                    </TableCell>
                  </TableRow>
                  <TableRow>
                    <TableCell align="center" id="infoTableTitle">
                      {this.props.iouType === 2 ? "Request" : "IOU"} Created on:
                    </TableCell>
                    <TableCell align="center" id="infoTableContent">
                      {this.props.created_time}
                    </TableCell>
                  </TableRow>
                  {this.props.details !== "" && (
                    <TableRow>
                      <TableCell align="center" id="infoTableTitle">
                        {this.props.iouType === 2
                          ? "Task:"
                          : this.props.iouType === 1
                          ? "Will be completed by:"
                          : "Will be completed by:"}
                      </TableCell>
                      <TableCell align="center" id="infoTableContent">
                        {this.props.iouType === 2
                          ? this.props.details
                          : this.props.iouType === 1
                          ? "Me :)"
                          : this.props.completed_by}
                      </TableCell>
                    </TableRow>
                  )}
                </TableBody>
              </Table>
            </TableContainer>
          </DialogContent>
          {this.props.iouType === 2 && (
            <DialogContent className="Content" id="completePopUpRewards">
              <DialogContentText variant="h6">{"Rewards"}</DialogContentText>
              {this.props.rewards.map((reward) =>
                this.renderPopUpRewards(reward)
              )}
            </DialogContent>
          )}
          {this.props.iouType !== 0 && (
            <DialogContent className="content">
              <DialogContentText variant="h6">
                {"Upload proof?"}
              </DialogContentText>
              <input
                type="file"
                onChange={(e) => {
                  if (e.target.files) {
                    this.setState({ submittedProof: e.target.files[0] });
                  }
                }}
                accept="image/*"
                id="inputProof"
              />
              <label htmlFor="inputProof">
                <Button variant="contained" color="primary" component="span">
                  {this.state.submittedProof ? "CHANGE" : "UPLOAD"}
                </Button>
              </label>
              {this.fileContent()}
            </DialogContent>
          )}
          <Divider />
          <DialogActions>
            <Button
              size="large"
              color="primary"
              onClick={() =>
                this.setState({ completeIOU: false, submittedProof: null })
              }
            >
              Nevermind
            </Button>
            <Button
              id="createRequest"
              size="large"
              color="primary"
              onClick={() => {
                if (this.props.iouType === 0) {
                  this.completeIouOwed();
                } else if (this.props.iouType === 1) {
                  this.completeIouOwe();
                } else {
                  this.completeRequest();
                }
                this.setState({
                  completeIOU: false,
                });
              }}
              autoFocus
              disabled={this.checkSubmitButton()}
            >
              {"Submit & Complete"}
            </Button>
          </DialogActions>
        </Dialog>
        <Snackbar
          anchorOrigin={{
            vertical: "bottom",
            horizontal: "left",
          }}
          message={this.state.snackMessage}
          open={this.state.completeSnack}
          onClose={() => {
            this.setState({ completeSnack: false });
          }}
          autoHideDuration={5000}
        />
      </div>
    );
  }
}

export default IouComplete;
