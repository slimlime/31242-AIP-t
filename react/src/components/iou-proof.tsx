import React from "react";
import "../assets/css/iou-request.css";
import { imageEndpoint } from "../api/endpoints";
import ImageIcon from "@material-ui/icons/Image";
import { Popover, Typography, Divider } from "@material-ui/core";

type IouProofProps = {
  imagePK: string | null;
  proof_of_debt: boolean;
};

type IouProofState = {
  resImage: string;
  anchorEl: HTMLElement | null;
};

class IouProof extends React.Component<IouProofProps, IouProofState> {
  state: IouProofState = {
    resImage: "",
    anchorEl: null,
  };

  render() {
    return (
      <div id="requestItem">
        {this.props.imagePK ? (
          <div>
            <div
              className="cursorPointer"
              onClick={(event: React.MouseEvent<HTMLElement>) => {
                this.setState({ anchorEl: event.currentTarget });
              }}
            >
              <ImageIcon color="primary" id="proofIcon" />
            </div>
            <Popover
              id="popUpMargin"
              open={Boolean(this.state.anchorEl)}
              anchorEl={this.state.anchorEl}
              onClose={() => this.setState({ anchorEl: null })}
              anchorOrigin={{
                vertical: "bottom",
                horizontal: "center",
              }}
              transformOrigin={{
                vertical: "top",
                horizontal: "center",
              }}
            >
              <Typography id="taskDetailPopUp">
                {this.props.proof_of_debt
                  ? "Proof of Debt"
                  : "Proof of Completion"}
                <Divider id="taskPopUpDivider" />
                <img
                  src={`${imageEndpoint}/${this.props.imagePK}`}
                  id="proofImage"
                  alt="Proof of Completion"
                />
              </Typography>
            </Popover>
          </div>
        ) : (
          <div>
            <div
              className="cursorPointer"
              onClick={(event: React.MouseEvent<HTMLElement>) => {
                this.setState({ anchorEl: event.currentTarget });
              }}
            >
              <ImageIcon id="noProofIcon" color="disabled" />
            </div>
            <Popover
              id="popUpMargin"
              open={Boolean(this.state.anchorEl)}
              anchorEl={this.state.anchorEl}
              onClose={() => this.setState({ anchorEl: null })}
              anchorOrigin={{
                vertical: "bottom",
                horizontal: "center",
              }}
              transformOrigin={{
                vertical: "top",
                horizontal: "center",
              }}
            >
              <Typography id="taskDetailPopUp">
                {this.props.proof_of_debt
                  ? "No proof of debt. This favour was created by the giver."
                  : "No proof of completion. This favour was completed by the receiver"}
              </Typography>
            </Popover>
          </div>
        )}
      </div>
    );
  }
}

export default IouProof;
