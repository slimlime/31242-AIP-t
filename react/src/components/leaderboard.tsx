import React from "react";
import "../assets/css/leaderboard.css";
import {
  leaderboardAllEndpoint,
  leaderboardMeEndpoint,
} from "../api/endpoints";
import {
  CircularProgress,
  Container,
  Divider,
  Grid,
  Paper,
  Typography,
} from "@material-ui/core";
import Pagination from "@material-ui/lab/Pagination";
import LeaderboardUser from "./leaderboardUser";
import { UserContext } from "./user-context";
import { Redirect } from "react-router-dom";

type LeaderboardUser = {
  rank: number;
  username: string;
  score: number;
};

type LeaderboardState = {
  users: LeaderboardUser[];
  me: {
    rank: number;
    score: number;
  };
  showMe: boolean;
  error: string;
  pageNumber: number;
  unauthRep: boolean;
};

type LeaderboardProps = {};

const numberOfItemsPerPage = 3;

class Leaderboard extends React.Component<LeaderboardProps, LeaderboardState> {
  private loadingRef: React.RefObject<HTMLInputElement>;

  constructor(props: LeaderboardProps) {
    super(props);

    this.loadingRef = React.createRef();
  }

  state: LeaderboardState = {
    users: [],
    error: "",
    me: { rank: -1, score: -1 },
    showMe: false,
    pageNumber: 1,
    unauthRep: false,
  };

  componentDidMount() {
    this.fetchAllLeaderboard();
    this.fetchMeLeaderboardIfNeeded();
  }

  static contextType = UserContext;

  // Determine whether user is logged in and fetch personal score if needed
  fetchMeLeaderboardIfNeeded = () => {
    if (this.context.user.name !== "?") {
      this.fetchMeLeaderboard();
    }
  };

  // Toggle loading circle for when async fetching
  setLoading(value: boolean): void {
    if (!this.loadingRef && !(this.loadingRef as any).current) {
      (this.loadingRef as any).current!.style.display = value
        ? "block"
        : "none";
    }
  }

  // Calculate the number of pages required to hold the users on the leaderboard with numberOfItemsPerPage per page
  setCountOfLeaderboard(): number {
    if (this.state.users.length % numberOfItemsPerPage !== 0) {
      return Math.ceil(this.state.users.length / numberOfItemsPerPage);
    } else {
      return this.state.users.length / numberOfItemsPerPage;
    }
  }

  // Fetch all the users on the leaderboard
  fetchAllLeaderboard() {
    this.setLoading(true);
    fetch(`${leaderboardAllEndpoint}`, {
      method: "GET",
    })
      .then((res) => {
        if (res.status === 200) {
          res.json().then((body) => {
            this.setState({ users: body });
            this.setState({ error: "" });
            this.setLoading(false);
          });
        } else if (res.status === 401) {
          this.setState({ unauthRep: true });
          this.setLoading(false);
        }
      })
      .catch((exception) => {
        console.error("Error:", exception);
        this.setState({ error: exception });
        this.setLoading(false);
      });
  }

  // Fetch personal score and rank for leaderboard
  fetchMeLeaderboard() {
    fetch(`${leaderboardMeEndpoint}`, {
      method: "GET",
    })
      .then((res) => {
        if (res.status === 200) {
          res.json().then((body) => {
            this.setState({ me: body }, () => {
              this.setState({ showMe: true });
            });
            this.setState({ error: "" });
            console.log("Success:", body);
          });
        } else if (res.status === 401) {
          this.setState({ unauthRep: true });
          this.setLoading(false);
        } else {
          res.json().then((body) => {
            console.error("Error:", body.errors);
            this.setState({ error: body.errors });
          });
        }
      })
      .catch((exception) => {
        console.log("Error:", exception);
      });
  }

  render() {
    // If unauthorised response (e.g. expired token), redirect to login
    if (this.state.unauthRep) {
      return (
        <Redirect
          to={{
            pathname: "/login",
            state: {
              unauthenticated:
                "Your session has expired! Please sign in again :)",
            },
          }}
        />
      );
    }

    return (
      <Paper elevation={3} className="content">
        <Container id="leaderboardContainer" component="main" maxWidth="lg">
          <Typography id="leaderboardHeader" component="h3" variant="h4">
            {"Leaderboard"}
          </Typography>
          <Grid
            container
            direction="column"
            justify="space-between"
            spacing={1}
          >
            <Grid item xs={12}>
              <Grid container direction="row">
                <Grid id="leaderboardItem" item xs={2}>
                  <Typography variant="subtitle1">Rank</Typography>
                </Grid>
                <Grid id="leaderboardItem" item xs={5}>
                  <Typography variant="subtitle1">Username</Typography>
                </Grid>
                <Grid id="leaderboardItem" item xs={2}>
                  <Typography variant="subtitle1">Score</Typography>
                </Grid>
                <Grid id="leaderboardItem" item xs={3}>
                  <Typography variant="subtitle1">Avatar</Typography>
                </Grid>
              </Grid>
            </Grid>
            <CircularProgress
              ref={this.loadingRef}
              size={35}
              color="inherit"
              id="leaderboardLoading"
            />
            {this.state.error}
            {/* kudos to
            https://stackoverflow.com/questions/61774099/how-to-add-pagination-to-a-long-list-using-react-materialui */}
            {this.state.users
              .slice(
                (this.state.pageNumber - 1) * numberOfItemsPerPage,
                this.state.pageNumber * numberOfItemsPerPage
              )
              .map((user, i) => {
                return (
                  <Grid item xs={12}>
                    <LeaderboardUser
                      key={i}
                      rank={user.rank}
                      username={user.username}
                      score={user.score}
                    />
                  </Grid>
                );
              })}
          </Grid>
          <Pagination
            id="pagination"
            count={this.setCountOfLeaderboard()}
            page={this.state.pageNumber}
            onChange={(event: React.ChangeEvent<unknown>, value: number) => {
              this.setState({ pageNumber: value });
            }}
            defaultPage={1}
            siblingCount={0}
            color="primary"
          />
          <Divider variant="middle" />
          {this.state.showMe && (
            <div id="meScore">
              <Typography variant="h6">
                And you are rank {this.state.me?.rank} with a score of{" "}
                {this.state.me?.score}!
              </Typography>
              <img
                alt="Raising hands"
                width="40"
                height="40"
                src={process.env.PUBLIC_URL + "/raising-hands.png"}
              />
              <img
                alt="Party popper"
                width="40"
                height="40"
                src={process.env.PUBLIC_URL + "/party-popper.png"}
              />
            </div>
          )}
        </Container>
      </Paper>
    );
  }
}

export default Leaderboard;
