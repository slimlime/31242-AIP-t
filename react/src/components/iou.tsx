import React from "react";
import IouComplete from "./iou-complete";
import IouProof from "./iou-proof";
import IouFavour from "./iou-single-favour";
import { Grid } from "@material-ui/core";
import { UserContext } from "../components/user-context";

type Item = {
  id: string;
  display_name: string;
};

type IouObj = {
  id: string;
  item: Item;
  giver: { username: string; display_name: string };
  receiver: { username: string; display_name: string };
  parent_request: string | null;
  proof_of_debt: string | null;
  proof_of_completion: string | null;
  created_time: string;
  claimed_time: string | null;
  is_claimed: boolean;
};

type IouProps = {
  iou: IouObj;
  iouType: number;
  refreshTable: () => void;
};

// Mainly a container component
class IOU extends React.Component<IouProps> {
  static contextType = UserContext;

  getCompletor(): string {
    if (this.props.iou.proof_of_completion !== null) {
      return this.props.iou.giver.username;
    } else {
      return this.props.iou.receiver?.username ?? "";
    }
  }

  getAuthor(): string {
    if (
      this.props.iou.proof_of_debt !== null &&
      this.props.iou.parent_request === null
    ) {
      return this.props.iou.receiver?.username ?? "";
    } else {
      return this.props.iou.giver.username;
    }
  }

  render() {
    return (
      <Grid container xs={12} spacing={1}>
        <Grid item xs={7} id="requestItemContainer">
          <IouFavour
            giverDisplayName={this.props.iou.giver.username}
            recieverDisplayName={this.props.iou.receiver?.username ?? "?"}
            item={this.props.iou.item}
          />
        </Grid>
        <Grid item xs={2} id="requestItemContainer">
          <IouProof
            imagePK={this.props.iou.proof_of_debt}
            proof_of_debt={true}
          />
          <IouProof
            imagePK={this.props.iou.proof_of_completion}
            proof_of_debt={false}
          />
        </Grid>
        <Grid item xs={3} id="requestProofContainer">
          <IouComplete
            id={this.props.iou.id}
            is_completed={this.props.iou.is_claimed}
            completed_by={this.getCompletor()}
            claimed_time={this.props.iou.claimed_time}
            created_time={this.props.iou.created_time}
            author={this.getAuthor()}
            rewards={[
              {
                id: "",
                giver: {
                  username: this.props.iou.giver.username,
                  display_name: this.props.iou.giver.username,
                },
                item: {
                  id: this.props.iou.item.id,
                  display_name: this.props.iou.item.display_name,
                },
              },
            ]}
            details=""
            iouType={this.props.iouType}
            refreshTable={this.props.refreshTable}
          />
        </Grid>
      </Grid>
    );
  }
}

export default IOU;
