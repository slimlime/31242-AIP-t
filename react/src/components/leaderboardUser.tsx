import React from "react";
import "../assets/css/leaderboardUser.css";
import { Avatar, Grid, Typography } from "@material-ui/core";
import { nameToUpperInitials } from "./avatarWithMenu";

type LeaderboardUserProps = {
  rank: number;
  username: string;
  score: number;
};

// Container component that holds rank, username and score in a horizontal grid
function LeaderboardUser(props: LeaderboardUserProps) {
  return (
    <Grid container direction="row">
      <Grid id="leaderboardItem" item xs={2}>
        <Typography>{props.rank}</Typography>
      </Grid>
      <Grid id="leaderboardItem" item xs={5}>
        <Typography>{props.username}</Typography>
      </Grid>
      <Grid id="leaderboardItem" item xs={2}>
        <Typography>{props.score}</Typography>
      </Grid>
      <Grid id="leaderboardItem" item xs={3}>
        <Avatar
          style={{
            justifyContent: "center",
            display: "flex",
            alignItems: "center",
          }}
          id="leaderboardUserAvatar"
        >
          {nameToUpperInitials(props.username)}
        </Avatar>
      </Grid>
    </Grid>
  );
}

export default LeaderboardUser;
