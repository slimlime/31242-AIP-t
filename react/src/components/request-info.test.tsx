import React from "react";
import { shallow, mount } from "enzyme";
import RequestInfo from "./request-info";
import { UserContext } from "../components/user-context";

describe("<RequestInfo />", () => {
  it("should render correctly", () => {
    const wrapper = shallow(
      <UserContext.Provider
        value={{
          user: { name: "Kevin Leung" },
          updateUser: () => {},
        }}
      >
        <RequestInfo
          request={{
            id: "1",
            author: { username: "James", display_name: "James Lee" },
            completed_by: { username: "Kevin", display_name: "Kevin Leung" },
            proof_of_completion: "Some ID",
            rewards: [
              {
                id: "1",
                giver: { username: "James", display_name: "James Lee" },
                item: { id: "202", display_name: "Hug" },
              },
              {
                id: "2",
                giver: { username: "Sean", display_name: "Sean Tran" },
                item: { id: "201", display_name: "Coffee" },
              },
            ],
            details: "Clean the fridge",
            created_time: "02/02/2020",
            completion_time: "02/02/2020",
            is_completed: true,
          }}
          refreshTable={() => {
            return "substitute test function";
          }}
        />
      </UserContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should render the modal if the icon is clicked", () => {
    const wrapper = mount(
      <UserContext.Provider
        value={{
          user: { name: "James" },
          updateUser: () => {},
        }}
      >
        <RequestInfo
          request={{
            id: "1",
            author: { username: "James", display_name: "James Lee" },
            completed_by: { username: "Kevin", display_name: "Kevin Leung" },
            proof_of_completion: "Some ID",
            rewards: [
              {
                id: "1",
                giver: { username: "James", display_name: "James Lee" },
                item: { id: "202", display_name: "Hug" },
              },
              {
                id: "2",
                giver: { username: "Sean", display_name: "Sean Tran" },
                item: { id: "201", display_name: "Coffee" },
              },
            ],
            details: "Clean the fridge",
            created_time: "02/02/2020",
            completion_time: "02/02/2020",
            is_completed: true,
          }}
          refreshTable={() => {
            return "substitute test function";
          }}
        />
      </UserContext.Provider>
    );
    wrapper.find("svg#infoIcon").simulate("click");
    expect(wrapper.state("infoModal")).toBe(true);
  });
});
