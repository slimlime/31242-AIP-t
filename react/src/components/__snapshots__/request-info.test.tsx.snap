// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestInfo /> should render correctly 1`] = `
<RequestInfo
  refreshTable={[Function]}
  request={
    Object {
      "author": Object {
        "display_name": "James Lee",
        "username": "James",
      },
      "completed_by": Object {
        "display_name": "Kevin Leung",
        "username": "Kevin",
      },
      "completion_time": "02/02/2020",
      "created_time": "02/02/2020",
      "details": "Clean the fridge",
      "id": "1",
      "is_completed": true,
      "proof_of_completion": "Some ID",
      "rewards": Array [
        Object {
          "giver": Object {
            "display_name": "James Lee",
            "username": "James",
          },
          "id": "1",
          "item": Object {
            "display_name": "Hug",
            "id": "202",
          },
        },
        Object {
          "giver": Object {
            "display_name": "Sean Tran",
            "username": "Sean",
          },
          "id": "2",
          "item": Object {
            "display_name": "Coffee",
            "id": "201",
          },
        },
      ],
    }
  }
/>
`;
